### PR TITLE
fix: [CDS-49774]: template propagation demo feedback changes

### PR DESCRIPTION
**Pull request by AshwinBhatkal-Harness**
### Summary

Template propagation demo feedback changes
1. Deleting first node should throw appropriate error - Handled as part of https://github.com/harness/harness-core-ui/pull/14393
2. Show previous stages for propagation that have not already been propagated
3. When on service tab of a normal stage, if navigated to a different stage template or another type of stage, default service schema gets appended. Fixed
4. Error strip was wrongly shown for empty service when useFromStage was selected
5. Negate FF check - silly miss

#### Screenshots

https://user-images.githubusercontent.com/96711906/223913995-b04e3901-e5a9-495c-811e-272841888ea8.mov

https://user-images.githubusercontent.com/96711906/223913997-505b1f54-a22c-4357-95b9-d60c1e09cf46.mov

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Sonar: `retrigger sonar`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Feature Name Check: `trigger featurenamecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress Rest: `retrigger cypress-rest`
- Cypress CD: `retrigger cypress-cd`
- Cypress Pipeline: `retrigger cypress-pipeline`
- Cypress CV: `retrigger cypress-cv`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

File: src/modules/70-pipeline/components/PipelineInputSetForm/ServicesInputSetForm/SingleServiceInputSetForm.tsx [hunk 1]
```diff
@@ -79,13 +80,18 @@ export default function SingleServiceInputSetForm({
   )
 
   const getStagesAllowedforPropogate = useCallback(
-    (stageItem): boolean => {
-      if (stageItem.stage.template) {
+    (stageItem: StageElementWrapperConfig): boolean => {
```
Comment (AshwinBhatkal-Harness) [inline src/modules/70-pipeline/components/PipelineInputSetForm/ServicesInputSetForm/SingleServiceInputSetForm.tsx:83]: Fixes which stages are shown for propagate. Stages propagating from previous stage are filtered

File: src/modules/75-cd/components/PipelineSteps/DeployServiceEntityStep/DeployServiceEntityStep.tsx [hunk 2]
```diff
@@ -134,11 +135,22 @@ export class DeployServiceEntityStep extends Step<DeployServiceEntityData> {
...
       set(errors, 'service.serviceRef', getString?.('cd.pipelineSteps.serviceTab.serviceIsRequired'))
     }
+
+    if (
+      isNil(data?.service?.serviceRef) &&
+      isEmpty(data?.service?.useFromStage?.stage) &&
+      isRequired &&
+      (isValueRuntimeInput(template?.service?.serviceRef) ||
+        isValueRuntimeInput(template?.service?.useFromStage as any))
+    ) {
```
Comment (AshwinBhatkal-Harness) [inline src/modules/75-cd/components/PipelineSteps/DeployServiceEntityStep/DeployServiceEntityStep.tsx:151]: Fixes unsaved changes for other stage/templates

File: src/modules/75-cd/components/PipelineStudio/DeployStageSetupShell/DeployStageSetupShell.tsx [hunk 3]
```diff
@@ -131,18 +131,24 @@ export default function DeployStageSetupShell(): JSX.Element {
...
   const setDefaultServiceSchema = useCallback((): Promise<void> => {
     const stageData = produce(selectedStage, draft => {
       if (draft) {
         if (isNewService) {
-          isEmpty(get(draft, 'stage.spec.service.serviceRef')) &&
+          if (
+            isEmpty(get(draft, 'stage.spec.service.serviceRef')) &&
             isEmpty(get(draft, 'stage.spec.services.values')) &&
             isEmpty(get(draft, 'stage.spec.service.useFromStage')) &&
+            // deploy type check - fixes weird issue which updates any other stage / stage template on unmount.
```
Comment (AshwinBhatkal-Harness) [inline src/modules/75-cd/components/PipelineStudio/DeployStageSetupShell/DeployStageSetupShell.tsx:143]: Fixes unsaved changes for other stage/templates

File: src/modules/75-cd/components/PipelineStudio/DeployServiceSpecifications/PropagateWidget/PropagateFromServiceV2.tsx [hunk 4]
```diff
@@ -32,15 +33,18 @@ export default function PropagateFromServiceV2({
...
   const formikRef = React.useRef<FormikProps<unknown> | null>(null)
   const { subscribeForm, unSubscribeForm } = React.useContext(StageErrorContext)
 
   useEffect(() => {
-    subscribeForm({ tab: DeployTabs.SERVICE, form: formikRef })
-    return () => unSubscribeForm({ tab: DeployTabs.SERVICE, form: formikRef })
+    if (subscribeToForm) {
+      subscribeForm({ tab: DeployTabs.SERVICE, form: formikRef })
+      return () => unSubscribeForm({ tab: DeployTabs.SERVICE, form: formikRef })
+    }
```
[the remark this answers is not in the document]
Comment (AshwinBhatkal-Harness) [reply src/modules/75-cd/components/PipelineStudio/DeployServiceSpecifications/PropagateWidget/PropagateFromServiceV2.tsx:47]: This useEffect is only for mount and unmount. Adding subscribeToForm as dependency can create confusion

**Review by vikrant-harness — DISMISSED** [review]

**Review by rameshx — DISMISSED** [review]

**Review by satyam-harness — DISMISSED** [review]

**Review by vikrant-harness — APPROVED** [review]

**Review by vkbansal-harness — APPROVED** [review]

**Pull request merged**